### PR TITLE
More intrusive_ptr

### PR DESCRIPTION
**Pull request by MaxKellermann**
The pieces start to fit together - these patches remove a lot of transitional boilerplate which had to be added before to make `IntrusivePtr<>` fit with existing code.

Let's see if those changes survive the CI tests (and how many rounds of corrections I'll need to do).

### CODE REVIEW

**Comment by timwoj** [comment]
Do all of the tests succeed on your local machine? It might be easier than going back and forth against CI.

**Comment by MaxKellermann** [comment]
> Do all of the tests succeed on your local machine? It might be easier than going back and forth against CI.

No, I didn't try that yet. I hacked those patches during a short brain idle (doing simple stuff like that can be very relaxing). Running the tests locally is cumbersome for me because I don't understand this "btest" thing yet, so just pushing it over here to let CirrusCI do the job for me was actually the easiest approach! But again, I was too optimistic. This old code base still has strange reference ownership semantics, but these are fading out.

**Comment by MaxKellermann** [comment]
Turns out d81bfed45da04d0e77496ea4487832d5aa1f95ba written by @rsmmr 6 years ago to fix a memory leak introduced a use-after-free bug, which was revealed by this PR and caused the CI failures.

**Comment by timwoj** [comment]
`btest` is basically integration testing for zeek. The majority of the tests are in the testing/btest directory, except for some long-running ones that live in external repos. The individual zeek files in the subdirectories under testing/btest describe how zeek runs for each test, if you need to run it manually to start it up in a debugger.

If you add the aux/btest directory to your path, you should have access to the `btest` binary. You can run it in parallel mode with the `-j` flag, just like make/ninja. You can also run btest with the `-d` flag to get more output for a failing tests. One big key is that a failed test creates a directory under testing/btest/.tmp that contains all of the scripts and output from the test.

There's more detail in testing/btest/README, but we really should have a wiki page about this.

**Comment by MaxKellermann** [comment]
3 use-after-free bugs introduced by d81bfed45da04d0e77496ea4487832d5aa1f95ba fixed by my new commit 72d99daa19d018d938eeb5010349416d807d5bec
I ran btest here, but a few tests still fail, which I also had with PRs that ran fine with CirrusCI (and were merged). So let's see what happens now.

**Comment by MaxKellermann** [comment]
Who knows what's up with the debian9 test failures?
The ubuntu16 build failure doesn't have enough context to see why it's really failing. (where is this template really instantiated?)
fedora30 is all green (hooray!), but memcheck finds new memory leaks - that will be easy to fix.

**Comment by jsiwek** [comment]
> Who knows what's up with the debian9 test failures?

The test itself looks invalid because the order of `zeek_init()` handler execution isn't well-defined unless a specific `&priority=N` is given.  Right thing to do is change the test such that they do each have a `&priority=N` attribute that will sort them in a well-defined order (can just go with what the order claimed by the current baseline).

> The ubuntu16 build failure doesn't have enough context to see why it's really failing. (where is this template really instantiated?)

You follow the "View more details on Cirrus CI" at the bottom through to the full context / build log:

https://cirrus-ci.com/task/5357924008591360

**Comment by MaxKellermann** [comment]
> The test itself looks invalid because the order of `zeek_init()` handler execution isn't well-defined unless a specific `&priority=N` is given.

Ah, a fragile test, I thought so.

> You follow the "View more details on Cirrus CI" at the bottom through to the full context / build log:

I did look at the full log, but the compiler didn't give enough context.
```
/zeek/src/IntrusivePtr.h:61:52: note: in instantiation of member function 'IntrusivePtr<Expr>::~IntrusivePtr' requested here
```
This is where the "instantiation requested here" chain ends, but it doesn't say where *this* template is instantiated. No time today, but when I take some time to analyze this, I'll try to compile Zeek with clang to check if my clang version reproduces this.

**Comment by MaxKellermann** [comment]
The memcheck failures are because `function_ingredients` now has proper ownership of reference-counted objects, and this old code:
https://github.com/zeek/zeek/blob/1b252038c548e3a1d4742294a3552eae120cf08c/src/Var.cc#L509-L512
But if I remove the `release()` line and let the `function_ingredients` object be destroyed, I get double-free bugs because the pointer to `function_ingredients::inits` is owned both by `function_ingredients` and `Stmt::inits` (via `BroFunc::AddInits()`).
Yet another attempt to fix reference counting foiled by old bugs.
I need to dig deeper to fix `end_func()` properly.

**Comment by MaxKellermann** [comment]
I played with this bug a bit, and I can now trigger the use-after-free bug with the following Zeek program:
```
function make_lambda(start : count) : function() : count
	{
	return function() : count
		{
			local dummy : string;
			return 0;
		};
	}

event zeek_init()
	{
	local one = make_lambda(1);
	local two = make_lambda(2);
	}
```
This program crashes with the `release/3.1` branch (so this has really nothing to do with my work). I did a fresh build of 3.1 and got this crash:
```
Thread 1 "zeek" received signal SIGSEGV, Segmentation fault.
0x0000555555b95d4b in Unref (o=0x1000600000007) at ../src/Obj.h:174
174		if ( o && --o->ref_cnt <= 0 )
(gdb) bt
#0  0x0000555555b95d4b in Unref (o=0x1000600000007) at ../src/Obj.h:174
#1  0x000055555611846e in InitStmt::~InitStmt (this=0x555557d7b140, __in_chrg=<optimized out>) at ../src/Stmt.cc:1664
#2  0x00005555561184c4 in InitStmt::~InitStmt (this=0x555557d7b140, __in_chrg=<optimized out>) at ../src/Stmt.cc:1667
#3  0x0000555555b95da6 in Unref (o=0x555557d7b140) at ../src/Obj.h:178
#4  0x0000555556117c6f in StmtList::~StmtList (this=0x5555581dde80, __in_chrg=<optimized out>) at ../src/Stmt.cc:1514
```
This program generates two instances of that lambda with a local, uninitialized variable. This `inits` list of uninitialized local variables is shared by two `BroFunc` instances (the two instantiated here), and is thus freed by both.
This happens at runtime, not at init time.

**Comment by MaxKellermann** [comment]
Lots of CI targets fail due to the fragile unit test.
memcheck fails with a memory leak.
But all the other problems (use-after-free crashes) are gone now!

**Comment by jsiwek** [comment]
> Lots of CI targets fail due to the fragile unit test.

There's this now in master if you'd like to bring it in to help de-noise: https://github.com/zeek/zeek/commit/363fc49518021b48e3981f44dbca9d8dda0407f1

> But all the other problems (use-after-free crashes and clang build failure) are gone now!

Nice, exciting that it's shaping up well.  I only skimmed the changes so far (mostly Val.h as it's most-traveled by external plugins); just trying to get an idea about potential breaking API changes.  For the most part looked like there were now either overloads that take either `IntrusivePtr` or raw pointer arguments, which is nice, but then also some methods are now changed to return `IntrusivePtr`, possibly breaking external code.  That may be fine, but just want to call attention in case there's strong opinions on how to handle.  My own suggestion:

* Just go ahead and break methods that change return type, make that clear in NEWS (we're already going this route for Zeek 3.2.0 in relation to some Opaque/HashVal methods)
* For methods that now have an additional overload for `IntrusivePtr` args, seems easy enough to keep the old one taking raw pointers around, but maybe want to add `[[deprecated("Remove in v4.1 ...")]]` to encourage migration and eventual removal.

**Comment by MaxKellermann** [comment]
>     * Just go ahead and break methods that change return type, make that clear in NEWS (we're already going this route for Zeek 3.2.0 in relation to some Opaque/HashVal methods)

This announcement already exists, I added it recently on your request:
https://github.com/zeek/zeek/blob/2286865674b0f9a792180eb5c0547b8c036c92d6/NEWS#L15-L18

>     * For methods that now have an additional overload for `IntrusivePtr` args, seems easy enough to keep the old one taking raw pointers around, but maybe want to add `[[deprecated("Remove in v4.1 ...")]]` to encourage migration and eventual removal.

I don't think that's worth it. My plan was to have these two overloads only during the transition phase, to make incremental migration to `IntrusivePtr` more seamless; the old variant shall be removed completely soon (and the new one rewritten to use `IntrusvePtr` properly). Having the old, unsafe one around also keeps unsafe code around. It is easy to see that all old code is unsafe and buggy!

File: src/Val.h [hunk 1]
```diff
@@ -936,18 +923,18 @@ class RecordVal : public Val, public notifier::Modifiable {
 
 class EnumVal : public Val {
 public:
-	Val* SizeVal() const override	{ return val_mgr->GetInt(val.int_val); }
+	EnumVal(EnumType* t, int i) : Val(bro_int_t(i), t)
+		{
+		}
```
[the remark this answers is not in the document]
Comment (MaxKellermann) [reply src/Val.h]: The problem is that a constructor used by `make_intrusive()` must also be accessible to that function. One solution is to make it `public`, the other is to `friend` the function, or just don't have this kind of "incremental" constructor around, and pass everything construction needs to the constructor.

File: src/Val.h [hunk 2]
```diff
@@ -956,20 +943,22 @@ class VectorVal : public Val, public notifier::Modifiable {
...
 	//
 	// Note: does NOT Ref() the element! Remember to do so unless
 	//       the element was just created and thus has refcount 1.
 	//
+	bool Assign(unsigned int index, IntrusivePtr<Val> element);
 	bool Assign(unsigned int index, Val* element);
-	bool Assign(Val* index, Val* element)
+
+	template<typename E>
+	bool Assign(Val* index, E &&element)
```
[the remark this answers is not in the document]
Comment (MaxKellermann) [reply src/Val.h]: Old habits! I fixed both.

**Comment by MaxKellermann** [comment]
All green now! Phew.

**Pull request merged**